File: java/src/uk/ac/ucl/ee/fnss/PropertyContainer.java
package uk.ac.ucl.ee.fnss;

import java.util.Hashtable;
import java.util.Map;
import java.util.Set;

/**
 * Generic container for properties
 *
 * @author Lorenzo Saino
 *
 */
public class PropertyContainer implements Cloneable {

	private Map<String, String> properties = new Hashtable<String, String>();

	/**
	 * Constructor
	 */
	public PropertyContainer() { }

	/**
	 * Return the value of a specific edge property
	 *
	 * @param name The name of the property
	 *
	 * @return The value of the property, if present. Otherwise, 
	 * <code>null</code> is returned.
	 */
	public String getProperty(String name) {
		return properties.get(name);
	}

	/**
	 * Set the value of a specific edge property. If a property with the same
	 * name already exists, it is overwritten
	 *
	 * @param name The name of the property
	 * @param value The value of the property
	 */
	public void setProperty(String name, String value) {
		properties.put(name, value);
	}

	/**
	 * Remove a specific property
	 * 
	 * @param name The name of the property to remove
	 * 
	 * @return the value of the property removed or <code>null</code> if no
	 * property with the given name was present. 
	 */
	public String removeProperty(String name) {
		return properties.remove(name);
	}
	
	/**
	 * Return a set containing all the property names. To get the value of a
	 * specific property use <code>getProperty(String name)</code>
	 * 
	 * @return a set with all property names of the object
	 * 
	 * @see #getProperty(String)
	 */
	public Set<String> getAllProperties() {
		return properties.keySet();
	}


	/**
	 * Return a copy of this object
	 * 
	 * @return A copy of this object
	 */
	@SuppressWarnings("unchecked")
	@Override
	public PropertyContainer clone() {
		PropertyContainer clone = null;
		try {
			clone = (PropertyContainer) super.clone();
		} catch (CloneNotSupportedException e) {
			/*
			 * It is safe to swallow the exception because Edge implements
			 * the cloneable interface, therefore this exception will never
			 * be thrown
			 */
			return null;
		}
		// deep copy property map otherwise this object and the clone
		// will point to the same properties object
		properties = (Map<String, String>)
				((Hashtable<String, String>) properties).clone();
		return clone;
	}

}

File: cpp/src/property-container.cpp
#include "property-container.h"

namespace fnss {

std::string PropertyContainer::getProperty(const std::string &name) const {
	propertiesType::const_iterator it;
	it = this->properties.find(name);

	if(it != this->properties.end())
		return it->second;
	else
		throw PropertyNotFoundException(name);
}

// std::string PropertyContainer::getProperty(const std::string &name,
// 										bool &found) const {
// 	propertiesType::const_iterator it;
// 	it = this->properties.find(name);
// 	if(it != this->properties.end()) {
// 		found = true;
// 		return it->second;
// 	}
// 	else {
// 		found = false;
// 		return std::string("");
// 	}
// }

void PropertyContainer::setProperty(const std::string &name,
								const std::string &value) {
	this->properties[name] = value;
}

bool PropertyContainer::hasProperty(const std::string &name) const {
	propertiesType::const_iterator it = this->properties.find(name);

	if(it == this->properties.end())
		return false;
	else
		return true;
}

void PropertyContainer::addProperties(const PropertyContainer &other) {
	propertiesType::const_iterator it;
	for(it = other.properties.begin(); it != other.properties.end(); it++)
		this->properties[it->first] = it->second;
}

std::string PropertyContainer::removeProperty(const std::string &name) {
	propertiesType::iterator it;
	it = this->properties.find(name);

	if(it != this->properties.end()) {
		std::string ret = it->second;
		this->properties.erase(it);
		return ret;
	} else
		throw PropertyNotFoundException(name);
}

// std::string PropertyContainer::removeProperty(const std::string &name,
// 												bool &found) {
// 	propertiesType::iterator it;
// 	it = this->properties.find(name);
//
// 	if(it != this->properties.end()) {
// 		found = true;
// 		std::string ret = it->second;
// 		this->properties.erase(it);
// 		return ret;
// 	} else {
// 		found = false;
// 		return std::string("");
// 	}
// }

std::set<std::string> PropertyContainer::getAllProperties() const {
	std::set<std::string> keys;
	propertiesType::const_iterator it;
	for(it = this->properties.begin(); it != this->properties.end(); it++)
		keys.insert(it->first);
	return keys;
}

}
